harvester_e2e_tests/integration/test_volumes.py

```diff
@@ -14,6 +15,95 @@
+    assert 201 == code, f"{msg}: {code}, {data}"
+
+    endtime = datetime.now() + timedelta(seconds=wait_timeout)
+    while endtime > datetime.now():
+        code, data = api_client.images.get(image_name)
+        if 'status' in data and 'progress' in data['status'] and data['status']['progress'] == 100:
+            break
+        sleep(sleep_timeout)
+    else:
+        raise AssertionError(f"{msg} ({wait_timeout}s timed out): {code}, {data}")
+
+    namespace = data['metadata']['namespace']
+    name = data['metadata']['name']
+    yield dict(ssh_user="ubuntu", id=f"{namespace}/{name}")
+
+    code, data = api_client.images.get(image_name)
+    if 200 == code:
+        code, data = api_client.images.delete(image_name)
+        # Wait until done so as not to affect futher TCs
+        endtime = datetime.now() + timedelta(seconds=wait_timeout)
+        while endtime > datetime.now():
+            code, data = api_client.images.get(image_name)
+            if 404 == code:
+                break
+            sleep(sleep_timeout)
+
+
+@pytest.fixture(scope="class")
+def focal_vm(api_client, unique_name, wait_timeout, sleep_timeout, focal_image):
+    vm_name = f"vm-focal-{unique_name}"
+    msg = "Failed to create VM"
+
+    vm_spec = api_client.vms.Spec(1, 2)
+    vm_spec.add_image(vm_name, focal_image["id"])
+    code, data = api_client.vms.create(vm_name, vm_spec)
+    assert 201 == code, f"{msg}: {code}, {data}"
+
+    endtime = datetime.now() + timedelta(seconds=wait_timeout)
+    while endtime > datetime.now():
+        code, data = api_client.vms.get(vm_name)
+        vm_status = data.get('status', {}).get('printableStatus', '')
+        if 200 == code and vm_status == "Running":
+            break
+        sleep(sleep_timeout)
+    else:
+        raise AssertionError(f"{msg} ({wait_timeout}s timed out): {code}, {data}")
+
+    volumes = list(filter(lambda vol: "persistentVolumeClaim" in vol,
+                          data["spec"]["template"]["spec"]["volumes"]))
+    assert len(volumes) == 1
+    yield data
+
+    code, data = api_client.vms.get(vm_name)
+    if 200 == code:
+        api_client.vms.delete(vm_name)
+        # Wait until done so as not to affect volume cleanup
+        endtime = datetime.now() + timedelta(seconds=wait_timeout)
+        while endtime > datetime.now():
+            code, data = api_client.vms.get(vm_name)
+            if 404 == code:
+                break
+            sleep(sleep_timeout)
+
+    vol_name = volumes[0]['persistentVolumeClaim']['claimName']
+    code, data = api_client.volumes.get(vol_name)
+    if 200 == code:
+        api_client.volumes.delete(vol_name)
+        # Wait until done so as not to affect image cleanup
+        endtime = datetime.now() + timedelta(seconds=wait_timeout)
+        while endtime > datetime.now():
+            code, data = api_client.volumes.get(vol_name)
+            if 404 == code:
+                break
+            sleep(sleep_timeout)
```

Comment:
Should we keep delete operation in try block? We might have test cases where we'll delete vms and volumes deleted as part of the test, in that case, will `api_client.vms.delete(vm_name)` and `api_client.volumes.delete(vol_name)` fail?

Reply:
Yes, TC may also have deletions, so I only **delete** if **get** returns 200.
But the delete-self still possible fail.
Just during refactoring, will also add a try block or rc verify on deletetion.

Reply:
Added RC check in latest commit